README.md
# XML Mockery

A little XML service that stands in for a vendor (and an example of a small Clojure API using [Liberator](https://clojure-liberator.github.io/liberator/))

## Usage

This API currently exposes only one endpoint, `/card_service`, that responds to POST requests sending XML request bodies, and returns XML responses. The vendor being replicated specifies its own request format, which is the following for most actions taken on a "card":

```xml
<?xml version="1.0" encoding="UTF‐8"?>
<TransferredValueTxn>
    <TransferredValueTxnReq> 
        <ReqCat>TransferredValue</ReqCat> 
        <ReqAction>Redeem</ReqAction> 
        <Date>20040327</Date> 
        <Time>221522</Time> 
        <PartnerName>MOCK</PartnerName>
        <CardActionInfo>
            <PIN>12340066</PIN>
            <AcctNum>MOCKCLIENT</AcctNum>
            <SrcRefNum>123412341234</SrcRefNum>
        </CardActionInfo>
    </TransferredValueTxnReq>
</TransferredValueTxn>
```

The actions the API knows how to respond to are (see the linked templates for the response format, anything in `{{double brackets}}` is variable data that the server will populate with sensible values):

* [Echo](https://github.com/lfborjas/mockery/blob/master/resources/card_responses/echo.xml) (returns whatever you send in the `EchoData` node)
* [StatInq](https://github.com/lfborjas/mockery/blob/master/resources/card_responses/card-action.xml)
* [Redeem](https://github.com/lfborjas/mockery/blob/master/resources/card_responses/card-action.xml)
* [RedeemReversal](https://github.com/lfborjas/mockery/blob/master/resources/card_responses/card-action.xml)


For `StatInq`, `Redeem` and `RedeemReversal`, if you send a `CardActionInfo/PIN` with the following use-case-specific **last 4 digits**, the response won't be success, but the specified message (notice that any suffix in the `:all` category will apply in any case, whereas the others depend on the `ReqAction` in the response):

```clj
  {:stat-inq        {:4001 "Card Is Active"
                     :4002 "Card Is Deactive"
                     :4003 "Card Is Redeemed"
                     :4006 "Card Not Found"
                     :4007 "Card Expired"}
   :redeem          {:0000 "Success"
                     :0043 "Card Is Invalid"
                     :0046 "Card Is Deactive"
                     :0038 "Card Is Redeemed"}
   :redeem-reversal {:0000 "Success"
                     :0067 "Not Reversible"}
   :all             {:0014 "Routing Error"
                     :0016 "Database Error"
                     :0019 "PIN Not found"
                     :0020 "Invalid PIN"
                     :0022 "MDN Not Found"
                     :0035 "Server Error"
                     :0036 "Server Error"
                     :0037 "Server Error"
                     :0082 "Invalid Identity Number"}}
```

For any of the three non-echo actions, sending the `PIN` `1234666` will return success, but it'll have different product info (useful if you want to see a different successful action).

For any action, sending a malformed XML request, specifying an unknown action, or failing to send a PIN for the card actions, a [bad request](https://github.com/lfborjas/mockery/blob/master/resources/card_responses/bad-request.xml) XML document will be returned.

Due to the semantics of POST, a status of `201 Created` will be returned in all cases when an XML request has been sent and content negotiation doesn't fail. The appropriate HTTP headers will be returned in any other case (e.g. if you send an `Accept` header of `text/html`, you'll receive a `406` response with a blank body).

Here's a poorly screenshot Postman session to illustrate a proper request and its response:

![image](https://user-images.githubusercontent.com/82133/46646039-e20d9600-cb55-11e8-981d-542dbb605f8a.png)

![image](https://user-images.githubusercontent.com/82133/46646049-f05bb200-cb55-11e8-969b-c3a778e7464c.png)



## Development

This project uses the [lein-ring](https://github.com/weavejester/lein-ring) plugin, which adds a couple of helpful Leiningen tasks:

* To run the server, run `lein ring server`
* To generate a standalone jar, run `lein ring uberjar`

### Deploying to Heroku

You'll have to make sure to [override the build environment variable](https://stackoverflow.com/questions/37962402/how-to-configure-build-command-in-herokus-clojure-buildpack): `heroku config:set LEIN_BUILD_TASK="ring uberjar"`--notice that the Procfile in this project relies on an uberjar generated with that particular command, the regular task (`uberjar`) will not generate the right `main` method!


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-10-07
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-10-07
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/mockery/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/mockery/compare/0.1.0...0.1.1


src/mockery/core.clj
(ns mockery.core
  (:require [liberator.core :refer [resource defresource]]
            [ring.middleware.params :refer [wrap-params]]
            [compojure.core :refer [defroutes ANY GET POST]]
            [mockery.services :as services]))

(defresource home [_]
  :available-media-types ["text/html"]
  :allowed-methods [:get]
  :handle-ok "<html>Hi there, please check my <a href=\"https://github.com/lfborjas/mockery/blob/master/README.md\">README</html>")

(defresource card-service [_]
  :available-media-types ["text/xml" "application/xml"]
  :allowed-methods [:post]
  :handle-created (fn [ctx]
                    (let [body (slurp (get-in ctx [:request :body]))]
                      (services/handle-card-request body))))

(defroutes app
  (ANY "/" [] home)
  (POST "/card_service" [] card-service))

(def handler
  (-> app
      wrap-params))


src/mockery/services.clj
(ns mockery.services
  (:require [clojure.xml :as xml]
            [clojure.data.xml :as data-xml]
            [clojure.zip :as zip]
            [clojure.data.zip.xml :as zip-xml]
            [mockery.views :as views]
            [mockery.models :as models]))

;; Inspired by:
;; http://blog.korny.info/2014/03/08/xml-for-fun-and-profit.html#parsing-wikipedia
;; http://clojure-doc.org/articles/tutorials/parsing_xml_with_zippers.html

(defn xml-zipper [str]
   (-> str
     data-xml/parse-str
     zip/xml-zip))


;; This convenience macro refactors away a bunch of very similar methods
;; taken from the xml zippers guide:
;; https://github.com/clojuredocs/guides/blob/master/articles/tutorials/parsing_xml_with_zippers.md
;; for example:
;; (macroexpand-1 '(get-text-at-path "TransferredValueTxn/TransferredValueTxnReq/EchoData" a))
;; returns
;; (clojure.data.zip.xml/xml1-> (mockery.services/xml-zipper a) :TransferredValueTxn :TransferredValueTxnReq :EchoData clojure.data.zip.xml/text)
;; that is, it takes the given document, turns it into a zipper
;; and traverses it until it gets the text of the last predicate

(defmacro get-text-at-path [path xml-doc]
  (let [segments (map keyword (clojure.string/split path #"/"))]
    `(zip-xml/xml1->
      (xml-zipper ~xml-doc)
      ~@segments
      zip-xml/text)))

(defn get-request-category [xml-req]
  (get-text-at-path
   "TransferredValueTxn/TransferredValueTxnReq/ReqCat"
   xml-req))

(defn get-request-action [xml-req]
  (get-text-at-path
   "TransferredValueTxn/TransferredValueTxnReq/ReqAction"
   xml-req))

(defn get-card-number [xml-req]
  (get-text-at-path
   "TransferredValueTxn/TransferredValueTxnReq/CardActionInfo/PIN"
   xml-req))

(defn get-echo-data [xml-req]
  (get-text-at-path
   "TransferredValueTxn/TransferredValueTxnReq/EchoData"
   xml-req))

(defn get-date [xml-req]
  (get-text-at-path
   "TransferredValueTxn/TransferredValueTxnReq/Date"
   xml-req))

(defn get-time [xml-req]
  (get-text-at-path
   "TransferredValueTxn/TransferredValueTxnReq/Time"
   xml-req))

(defn get-account [xml-req]
  (get-text-at-path
   "TransferredValueTxn/TransferredValueTxnReq/CardActionInfo/AcctNum"
   xml-req))

(defn get-ref [xml-req]
  (get-text-at-path
   "TransferredValueTxn/TransferredValueTxnReq/CardActionInfo/SrcRefNum"
   xml-req))

(defn get-tv-action [xml-req]
  (if
   (= "TransferredValue" (get-request-category xml-req))
   (get-request-action xml-req)))

;; takes a clojure symbol and returns a CamelCased string
;; e.g. (case->action :status-inq)
;; => StatusInq
(defn camelize [use-case]
  (-> use-case
      name
      (#(clojure.string/split % #"-"))
      (#(map  clojure.string/capitalize %))
      (#(clojure.string/join %))))

(defn respond [use-case card-number xml-req]
  (let [card-data   (models/match-card-number
                      use-case card-number)
        client-date (get-date xml-req)
        client-time (get-time xml-req)
        account     (get-account xml-req)
        ref-num     (get-ref xml-req)
        action-name (camelize use-case)]
    (views/render-card-response
     "card-action"
     (merge card-data
            {:pin card-number
             :acct-num account
             :src-ref-num ref-num
             :action action-name
             :given-date client-date
             :given-time client-time}))))

(defn bad-request [xml-req]
  (let [client-date (get-date xml-req)
        client-time (get-time xml-req)]
    (views/render-card-response
     "bad-request"
     {:given-date client-date
      :given-time client-time})))

(defn card-action [use-case xml-req]
  (if-let [card-number (get-card-number xml-req)]
    (respond use-case card-number xml-req)
    (bad-request xml-req)))

(defn echo [xml-req]
  (let [echo-data (get-echo-data xml-req)
        client-date (get-date xml-req)
        client-time (get-time xml-req)]
    (views/render-card-response "echo"
                                {:echo-data echo-data
                                 :given-date client-date
                                 :given-time client-time})))

(defn handle-card-request [xml-req]
  (cond
   (= "Redeem" (get-tv-action xml-req))
   (card-action :redeem xml-req)
   (= "RedeemReversal" (get-tv-action xml-req))
   (card-action :redeem-reversal xml-req)
   (= "StatInq" (get-tv-action xml-req))
   (card-action :stat-inq xml-req)
   (= "Echo" (get-tv-action xml-req))
   (echo xml-req)
   :else (bad-request xml-req)))


src/mockery/models.clj
(ns mockery.models)

(def status-codes
  {:stat-inq        {:4001 "Card Is Active"
                     :4002 "Card Is Deactive"
                     :4003 "Card Is Redeemed"
                     :4006 "Card Not Found"
                     :4007 "Card Expired"}
   :redeem          {:0000 "Success"
                     :0043 "Card Is Invalid"
                     :0046 "Card Is Deactive"
                     :0038 "Card Is Redeemed"}
   :redeem-reversal {:0000 "Success"
                     :0067 "Not Reversible"}
   :all             {:0014 "Routing Error"
                     :0016 "Database Error"
                     :0019 "PIN Not found"
                     :0020 "Invalid PIN"
                     :0022 "MDN Not Found"
                     :0035 "Server Error"
                     :0036 "Server Error"
                     :0037 "Server Error"
                     :0082 "Invalid Identity Number"}})

(def cards
  {:default {:product-face-value 30.0
             :product-upc "123456789"
             :activation-date "2018-10-01 22:00:00"
             :serial-number "987654321"}
   :1234666 {:product-face-value 66.6
             :product-upc "312654987"
             :activation-date "2017-10-31 03:33:33"
             :serial-number "918273465"}})

(defn match-status-code [use-case card-number]
  (let [conditions (merge (:all status-codes)
                          (use-case status-codes))]
    (first (filter
            (fn [[code _]]
              (clojure.string/ends-with? card-number
                                         (name code)))
            conditions))))

(defn match-card-number [use-case card-number]
  (let [status-data (if-let [sc (match-status-code use-case card-number)]
                      (zipmap [:resp-code :resp-msg] sc)
                      {:resp-code :0000 :resp-msg "Success"})
        card-data (get cards (keyword card-number) (:default cards))]
    (merge status-data card-data)))
